cache/lru_cache.cc

```diff
@@ -430,30 +433,62 @@ void LRUCacheShard::Promote(LRUHandle* e) {
 
   assert(secondary_handle->IsReady());
   e->SetIncomplete(false);
-  e->SetInCache(true);
   e->value = secondary_handle->Value();
-  e->CalcTotalCharge(secondary_handle->Size(), metadata_charge_policy_);
+  size_t secondary_handle_size = secondary_handle->Size();
+  e->CalcTotalCharge(secondary_handle_size, metadata_charge_policy_);
   delete secondary_handle;
 
   // This call could fail if the cache is over capacity and
   // strict_capacity_limit_ is true. In such a case, we don't want
   // InsertItem() to free the handle, since the item is already in memory
-  // and the caller will most likely just read from disk if we erase it here.
+  // and the caller will most likely just read from disk if we erase it
+  // here.
   if (e->value) {
-    Cache::Handle* handle = reinterpret_cast<Cache::Handle*>(e);
-    Status s = InsertItem(e, &handle, /*free_handle_on_fail=*/false);
+    bool insert_dummy_handle{false};
+    {
+      // Insert a dummy handle and return a standalone handle to caller
+      // when secondary_cache_ is CompressedSecondaryCache and e is a
+      // standalone handle.
+      if (use_compressed_secondary_cache_ && e->IsStandalone()) {
+        insert_dummy_handle = true;
+        // Update the properties for the standalone handle.
+        e->SetInCache(false);
+      } else {
+        e->SetInCache(true);
+        e->SetIsStandalone(false);
+      }
+    }
+
+    Status s;
+    if (insert_dummy_handle) {
+      // Insert a dummy handle into the primary cache and the charge is the
+      // size of the secondary handle. This dummy handle is
+      // not IsSecondaryCacheCompatible().
+      Cache::Priority priority =
+          e->IsHighPri() ? Cache::Priority::HIGH : Cache::Priority::LOW;
+      s = Insert(e->key(), e->hash, /*value=*/nullptr, secondary_handle_size,
+                 /*deleter=*/nullptr, /*helper=*/nullptr, /*handle=*/nullptr,
```

Comment:
We don't hold the handle here and wait it to be naturally evicted. So it is possible that the charge is removed before the block itself. Is it correct? Also, do we remove the charge of the secondary handle size? Can we charge the handle together with the block itself?

Reply:
I can charge the standalone handle itself, although it is not inserted into the primary cache.